### PR TITLE
Lw/optimizations

### PR DESCRIPTION
**Pull request by phi-lira**
### Purpose of this PR
 Reduces about 40MB of bandwidth cost in BoatAttack. (mostly store)

 - LWRP doesn't support camera stacking. changed Camera Clear Flags to use Background Type. The following options are available: Skybox, Solid Color, Don't Initialize. 
   On mobile Skybox and Don't Initialize are setup with DontCare load action. 
   Solid Color is setup with Clear. 
   Camera Clear Flags DepthOnly is replaced with SolidColor.

camerainspector

 - When supported. ARGB2101010 is used as HDR format in mobile. We were relying on DefaultHDR to pick the value from GraphicsTier setttings by this does not seem the case. Ideally we want to expose this in the pipeline quality settings.
 - Added support for non-square shadow map atlas. This saves half memory and bandwidth when rendering 2 directional shadow cascades.
 - Depth Texture is only created when necessary. We use depth RenderBuffer whenever possible.
 - Avoids extra uneccessary blit when rendering to Offscreen cameras.
 - Combine RenderOpaques and Skybox pass whenever there's not a custom render pass in between. 

---
### Release Notes
 Added

---
### Testing status
**Katana Tests**: https://katana.bf.unity3d.com/projects/com.unity.render-pipelines/builders?automation-tools_branch=add-platform-filter&ScriptableRenderLoop_branch=lw%2Foptimizations&unity_branch=trunk

**Manual Tests**: 
- Tested manually both single and stereo rendering (Mock HMD) using the LWGraphicsTest. 
- Tested bandwidth savings with BoatAttack.

**Automated Tests**: No automated test added. Had to disable test 015 as it depended on CameraStacking. It was working by chance. We need to update the test to inject a custom render pass to render the gun camera.

Any test projects to go with this to help reviewers? LWGraphicsTests

---
### Overall Product Risks
**Technical Risk**:  Medium. Considerable changes in different systems.

**Halo Effect**: Medium. Changes render target setup, this affects all platforms, VR included.

---
### Comments to reviewers
 - Final Blit check should only be performance when rendering to backbuffer. The checks for Display and kill alpha are not considered when rendering offscreen cameras.
 - Don't create a depth texture unless we can copy it to use as input for another pass. Changed SetRenderTarget method in ScriptableRenderPass to check if we should use render buffer. All RenderPasses are using this utilitary SetRenderTarget method.

### CODE REVIEW

File: com.unity.render-pipelines.lightweight/Runtime/ScriptableRenderer.cs [hunk 1]
```diff
@@ -301,8 +301,12 @@ public static RenderTextureDescriptor CreateRenderTextureDescriptor(ref CameraDa
             {
                 desc = new RenderTextureDescriptor(camera.pixelWidth, camera.pixelHeight);
             }
-            desc.colorFormat = cameraData.isHdrEnabled ? RenderTextureFormat.DefaultHDR :
-                RenderTextureFormat.Default;
+            
+            bool useHDR32 = Application.isMobilePlatform &&
```
Comment (aleks01010101) [inline com.unity.render-pipelines.lightweight/Runtime/ScriptableRenderer.cs]: Perhaps better to call this useRGB10A2? HDR32 sounds a bit misleading :)

File: com.unity.render-pipelines.lightweight/Runtime/Passes/MainLightShadowCasterPass.cs [hunk 2]
```diff
@@ -180,18 +186,19 @@ void SetupMainLightShadowReceiverConstants(CommandBuffer cmd, ref ShadowData sha
...
+            for (int i = 0; i < cascadeCount; ++i)
+                m_MainLightShadowMatrices[i] = m_CascadeSlices[i].shadowTransform;
 
             // We setup and additional a no-op WorldToShadow matrix in the last index
             // because the ComputeCascadeIndex function in Shadows.hlsl can return an index
             // out of bounds. (position not inside any cascade) and we want to avoid branching
             Matrix4x4 noOpShadowMatrix = Matrix4x4.zero;
-            noOpShadowMatrix.m33 = (SystemInfo.usesReversedZBuffer) ? 1.0f : 0.0f;
-            m_MainLightShadowMatrices[k_MaxCascades] = noOpShadowMatrix;
+            noOpShadowMatrix.m22 = (SystemInfo.usesReversedZBuffer) ? 1.0f : 0.0f;
```
[the remark this answers is not in the document]
Comment (phi-lira) [reply com.unity.render-pipelines.lightweight/Runtime/Passes/MainLightShadowCasterPass.cs]: That was intentional. Before it was a 0-base index stupid mistake.

File: com.unity.render-pipelines.lightweight/Runtime/ForwardRendererSetup.cs [hunk 3]
```diff
@@ -228,14 +240,20 @@ public void Setup(ScriptableRenderer renderer, ref RenderingData renderingData)
 
             if (camera.clearFlags == CameraClearFlags.Skybox)
             {
-                m_DrawSkyboxPass.Setup(colorHandle, depthHandle);
+                // We can't combine skybox and render opaques pases if there's a custom render pass in between
```
[the remark this answers is not in the document]
Comment (phi-lira) [reply com.unity.render-pipelines.lightweight/Runtime/ForwardRendererSetup.cs]: Fixed

File: com.unity.render-pipelines.lightweight/Runtime/ScriptableRenderer.cs (hunk 1 above)
Comment (phi-lira) [reply com.unity.render-pipelines.lightweight/Runtime/ScriptableRenderer.cs]: Fixed

File: com.unity.render-pipelines.lightweight/CHANGELOG.md [hunk 4]
```diff
@@ -10,9 +10,13 @@ and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.
 - You can now inject a custom render pass before rendering opaque objects. To do so, implement an `IBeforeRender` interface.
 
 ### Changed
+- You can now only initialize a camera by settings a Background Type. Skybox, Solid Color and Don't Initialize are the supported options.
```
Comment (aleks01010101) [inline com.unity.render-pipelines.lightweight/CHANGELOG.md]: by settings -> by setting?

File: com.unity.render-pipelines.lightweight/Editor/LightweightRenderPipelineCameraEditor.cs [hunk 5]
```diff
@@ -25,6 +33,15 @@ internal class Styles
...
             public readonly string mssaDisabledWarning = "Anti-aliasing is disabled in the Lightweight Render Pipeline asset.";
 
+            public static GUIContent[] cameraBackgroundType =
+            {
+                new GUIContent("Skybox"),
+                new GUIContent("Solid Color"),
+                new GUIContent("Don't Initialize"),
+            };
+
+            public static int[] cameraBackgroundValues = { 0, 1, 2};
```
Comment (aleks01010101) [inline com.unity.render-pipelines.lightweight/Editor/LightweightRenderPipelineCameraEditor.cs]: extra space: "{ 0"

File: com.unity.render-pipelines.lightweight/Editor/LightweightRenderPipelineCameraEditor.cs [hunk 6]
```diff
@@ -160,6 +176,52 @@ public override void OnInspectorGUI()
...
+            // Converts between ClearFlags and Background Type.
+            BackgroundType backgroundType = GetBackgroundType((CameraClearFlags) settings.clearFlags.intValue);
+
+            EditorGUI.BeginChangeCheck();
+            BackgroundType selectedType = (BackgroundType)EditorGUILayout.IntPopup(Styles.backgroundType, (int)backgroundType,
+                Styles.cameraBackgroundType, Styles.cameraBackgroundValues);
+
+            if (EditorGUI.EndChangeCheck())
+            {
+                CameraClearFlags selectedClearFlags;
```
Comment (aleks01010101) [inline com.unity.render-pipelines.lightweight/Editor/LightweightRenderPipelineCameraEditor.cs]: I'd move this switch to a separate function as you did with the conversion the other way around.

File: com.unity.render-pipelines.lightweight/Runtime/ScriptableRenderer.cs [hunk 7]
```diff
@@ -301,12 +301,18 @@ public static RenderTextureDescriptor CreateRenderTextureDescriptor(ref CameraDa
...
                 desc = new RenderTextureDescriptor(camera.pixelWidth, camera.pixelHeight);
             }
-            desc.colorFormat = cameraData.isHdrEnabled ? RenderTextureFormat.DefaultHDR :
-                RenderTextureFormat.Default;
-            desc.enableRandomWrite = false;
-            desc.sRGB = true;
+            
+            bool useRGB10A2 = Application.isMobilePlatform &&
+             SystemInfo.SupportsRenderTextureFormat(RenderTextureFormat.ARGB2101010);
+            RenderTextureFormat hdrFormat = (useRGB10A2) ? RenderTextureFormat.ARGB2101010 : RenderTextureFormat.DefaultHDR;
```
Comment (aleks01010101) [inline com.unity.render-pipelines.lightweight/Runtime/ScriptableRenderer.cs]: and some more nitpicking: this line uses a form (bla) ? bla : bla, next line uses bla ? bla : bla

File: com.unity.render-pipelines.lightweight/Runtime/ScriptableRenderer.cs [hunk 8]
```diff
@@ -315,16 +321,16 @@ public static ClearFlag GetCameraClearFlag(Camera camera)
...
-            CameraClearFlags cameraClearFlags = camera.clearFlags;
-            if (cameraClearFlags != CameraClearFlags.Nothing)
-            {
-                clearFlag |= ClearFlag.Depth;
-                if (cameraClearFlags == CameraClearFlags.Color || cameraClearFlags == CameraClearFlags.Skybox)
-                    clearFlag |= ClearFlag.Color;
-            }
+            // LWRP doesn't support CameraClearFlags.DepthOnly.
+            // In case of skybox we know all pixels will be rendered to screen so
+            // we don't clear color. In Vulkan/Metal this becomes DontCare load action
```
Comment (aleks01010101) [inline com.unity.render-pipelines.lightweight/Runtime/ScriptableRenderer.cs]: GL as well, I suppose (done via invalidate/clear).

File: com.unity.render-pipelines.lightweight/Runtime/ScriptableRenderer.cs [hunk 9]
```diff
@@ -315,16 +321,16 @@ public static ClearFlag GetCameraClearFlag(Camera camera)
...
+            if ((camera.clearFlags == CameraClearFlags.Skybox && RenderSettings.skybox != null) ||
+                 camera.clearFlags == CameraClearFlags.Nothing)
```
Comment (aleks01010101) [inline com.unity.render-pipelines.lightweight/Runtime/ScriptableRenderer.cs]: Should it really clear depth if the flag is set to "clear nothing"?

File: com.unity.render-pipelines.lightweight/CHANGELOG.md (hunk 4 above)
Comment (Katstodian) [reply com.unity.render-pipelines.lightweight/CHANGELOG.md]: Good catch. 
I've changed the sentence around and done some other edits as well. :)

File: com.unity.render-pipelines.lightweight/Runtime/ScriptableRenderer.cs (hunk 8 above)
Comment (phi-lira) [reply com.unity.render-pipelines.lightweight/Runtime/ScriptableRenderer.cs]: It's commented just below about the glinvalidate. I could have merged this comments indeed.

File: com.unity.render-pipelines.lightweight/Runtime/ScriptableRenderer.cs (hunk 7 above)
[the remark this answers is not in the document]
Comment (phi-lira) [reply com.unity.render-pipelines.lightweight/Runtime/ScriptableRenderer.cs]: We can add option in the pipeline asset. I'd rather do this in this PR as it affects performance and we are running perf tests soonish. When we have another solution we update this with that.

**Pull request merged**